pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,352 @@
+            List of (N^2)x(N^2) numpy complex arrays and corresponding paulis words
+        """
+
+        operators = []
+        names = []
+        # get all possible combinations of paulis
+        pauli_combinations = it.product(
+            list(range(len(self.pl_paulis))), repeat=self.nqubits
+        )
+        # remove the identity
+        next(pauli_combinations)
+        # construct the corresponding pennylane observables
+        for ps in pauli_combinations:
+            operators.append(self.get_pauli_op(ps))
+            names.append(ps)
+        return operators, [
+            "".join([self.pauli_int_to_str[i] for i in n]) for n in names
+        ]
+
+    def get_pauli_op(self, paulistring):
+        r"""Create a qml.Observable for a String corresponding to a Pauli word.
+
+        Args:
+            paulistring: String of the form 'IXYI...Z' corresponding to a Pauli word
+
+        Returns:
+            qml.Observable
+        """
+        for idx, p in enumerate(paulistring):
+            if p != 0:
+                op = self.pl_paulis[p](idx)
+                for q in paulistring[idx + 1:]:
+                    idx += 1
+                    if q != 0:
+                        op @= self.pl_paulis[q](idx)
+                break
+        return op
+
+    def get_omegas(self):
+        r"""Measure the coefficients of the Lie gradient with respect to a Pauli word basis
+
+        We want to calculate the components of the Lie gradient with respect to a Pauli word basis
+        For a Hamiltonian of the form :math:`H = \sum_i O_i`, this can be achieved by calculating
+
+        .. math:
+
+            \omega_{i,j} = \text{Tr}([\rho, O_i] P_j)
+
+        where :math:`P_j` is a Pauli word in the set of Pauli monomials on :math:`N` qubits.
+
+        Via the parameter shift rule, the commutator can be calculated as
+
+        .. math:
+
+            [\rho, O_i] = \frac{1}{2}(V(\pi/2) \rho V^\dag(\pi/2) - V(-\pi/2) \rho V^\dag(-\pi/2))
+
+        where :math:`V` is the unitary generated by the Pauli word :math:`V(\theta) = \exp{-i\theta P_j }`.
+
+        Returns:
+            Array of omegas for each direction in the Lie algebra.
+
+        """
+
+        obs_groupings, _ = qml.grouping.group_observables(
+            self.observables, self.coeffs
+        )
+        # get all circuits we need to calculate the coefficients
+        circuits = self.algebra_commutator(self.circuit.qtape, obs_groupings)[0]
+        # For each observable O_i in the Hamiltonian, we have to calculate all Lie coefficients
+        omegas = np.zeros((len(self.coeffs), len(self.lie_algebra_basis_names)))
+        idx = 0
+        for circuit_plus, circuit_min in circuits:
+            out_plus = qml.execute(
+                circuit_plus, self.circuit.device, gradient_fn=qml.gradients.param_shift
+            )
+            out_min = qml.execute(
+                circuit_min, self.circuit.device, gradient_fn=qml.gradients.param_shift
+            )
```

Comment:
@therooler you could optimize this further by executing all circuits at once,

```python
out = qml.execute(circuits, ...)
for out_plus, out_min in out:
    ...
```

Reply:
Hmm this
```python
out = qml.execute(
                circuits, self.circuit.device, gradient_fn=None
        )
        for out_plus, out_min in out:
            # depending on the length of the grouped observable, store the omegas in the array
            omegas[idx : idx + len(out_plus[0]), :] = 0.5 * (
                np.array(out_plus).T - np.array(out_min).T
            )
            idx += len(out_plus[0])
        return np.dot(self.coeffs, omegas)
```

throws an error:

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../pennylane/optimize/lie_gradient.py:246: in step
    self.step_and_cost(
../../pennylane/optimize/lie_gradient.py:197: in step_and_cost
    omegas = self.get_omegas()
../../pennylane/optimize/lie_gradient.py:321: in get_omegas
    out = qml.execute(
../../pennylane/interfaces/batch/__init__.py:313: in execute
    with qml.tape.Unwrap(*tapes):
../../pennylane/tape/unwrap.py:82: in __enter__
    stack.enter_context(
../../../../anaconda3/envs/pennylane/lib/python3.8/contextlib.py:425: in enter_context
    result = _cm_type.__enter__(cm)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pennylane.tape.unwrap.UnwrapTape object at 0x7f7ac8ed0b20>

    def __enter__(self):
>       self._original_params = self.tape.get_parameters(trainable_only=False)
E       AttributeError: 'tuple' object has no attribute 'get_parameters'

../../pennylane/tape/unwrap.py:132: AttributeError

`

Reply:
Ah my mistake 🤦 This is because the `qml.execute` only accepts a flat list of tapes. You would need to do something like this:

```python
c_plus, c_minus = list(zip(*circuits))

out = qml.execute(c_plus + c_minus, ...)
o_plus = out[:len(out) // 2]
o_minus = out[len(out) // 2:]

for out_plus, out_min in zip(o_plus, o_minus):
    ...
```

The main advantage here is that you are only calling `execute` once, so if you are using a cloud-based device, only one job is submitted over the API.

Reply:
Hmm I'm running into some issues here because the returned circuit outputs have different lengths based on the Pauli groupings. Should be able to fix this...